bundle/edu.gemini.catalog/src/main/scala/edu/gemini/catalog/votable/VoTableParser.scala

```diff
@@ -48,53 +49,114 @@ object VoTableParser extends VoTableParser {
   /**
    * parse takes an input stream and attempts to read the xml content and convert it to a VoTable resource
    */
-  def parse(url: String, is: InputStream): CatalogResult =
+  def parse(url: URL, is: InputStream): CatalogResult =
     validate(is).fold(k => \/.left(ValidationError(url)), r => \/.right(parse(XML.loadString(r))))
 }
 
-// A MagnitudesFilter can ignore fields for certain catalogues and transform others
-sealed trait MagnitudesFilter {
-  // Indicates if a field should be ignored
-  def ignoredMagnitudeField(v: FieldId): Boolean = false
-  // Indicates if a magnitude is valid
-  def validMagnitude(m: Magnitude): Boolean = !(m.value.isNaN || m.error.exists(_.isNaN))
+// A CatalogAdapter improves parsing handling catalog-specific options like parsing magnitudes and selecting key fields
+sealed trait CatalogAdapter {
+  // Required fields
+  val idField: FieldId
+  val raField: FieldId
+  val decField: FieldId
+  val pmRaField = FieldId("pmra", VoTableParser.UCD_PMRA)
+  val pmDecField = FieldId("pmde", VoTableParser.UCD_PMDEC)
+
+  // Indicates if the field is a magnitude
+  def isMagnitudeField(v: (FieldId, String)): Boolean = containsMagnitude(v._1) && !v._1.ucd.includes(VoTableParser.STAT_ERR) && v._2.nonEmpty
+  // Indicates if the field is a magnitude error
+  def isMagnitudeErrorField(v: (FieldId, String)): Boolean = containsMagnitude(v._1) && v._1.ucd.includes(VoTableParser.STAT_ERR) && v._2.nonEmpty
+  // filter magnitudes as a whole, removing invalid values and duplicates
+  def filterAndDeduplicateMagnitudes(magnitudeFields: List[(FieldId, Magnitude)]): List[Magnitude] = magnitudeFields.collect { case (_, mag) if validMagnitude(mag) => mag }
+  // Attempts to extract a band and value for a magnitude from a pair of field and value
+  def parseMagnitude(p: (FieldId, String)): CatalogProblem \/ (FieldId, MagnitudeBand, Double) = {
+    val (fieldId: FieldId, value: String) = p
+
+    val band = fieldToBand(fieldId)
+
+    for {
+      b <- band \/> UnmatchedField(fieldId.ucd)
```

Comment:
@sraaphorst yeah Eugene's tutorial is pretty good (but slightly out of date I think). You can also just ask me, or get on the `#scala` IRC channel on FreeNode and ask questions.